### PR TITLE
Metamorphosis - FoodSequence 3

### PR DESCRIPTION
**Pull request by TheShuEd**
## About the PR
specific recipes in a FoodSequence can turn a food into a separate special entity. The same result can be obtained in different ways. Different ways affect the final food solution, its flavors, and the garbage that remains after it is eaten.

for example, Cheeseburger recipe:
```yml
- type: metamorphRecipe
  id: FoodBurgerCheese
  key: burger
  result: FoodBurgerCheese
  rules:
  - !type:SequenceLength
    range:
      min: 3
      max: 4
  - !type:LastElementHasTags
    tags:
    - Bun
  - !type:IngredientsWithTags
    tags:
    - Cooked
    - Meat
    count:
      min: 1
      max: 2
  - !type:IngredientsWithTags
    tags:
    - Cheese
    count:
      min: 1
      max: 2
```

https://github.com/user-attachments/assets/bc9132a7-fdbc-4dde-929c-f3971be03e74

## Requirements

- [x] I have read and I am following the [Pull Request Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html). I understand that not doing so may get my pr closed at maintainer’s discretion
- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
all FoodSequenceElement components edited again
transfomed from
```yml
- type: FoodSequenceElement
    sprite:
      sprite: Objects/Consumable/Food/xxx.rsi
      state: xxx
    entries:
      burger:
        name: food-sequence-content-xxx
      taco: 
        name: food-sequence-content-xxx
```
to 
```yml
- type: FoodSequenceElement
    entries:
      Taco: xxx
      Burger: xxx
```
with new Dictionary `<ProtoId<Tag>, ProtoId<FoodSequenceElementPrototype>>`
new FoodSequenceElementPrototype required:

```yml
- type: foodSequenceElement
  id: Berries
  name: food-sequence-content-berries
  sprites:
  - sprite: Objects/Specific/Hydroponics/berries.rsi
    state: produce
  tags:
  - Fruit
```

**Changelog**
:cl:
- add: Returned Taco microwave recipes (people were sad)
- add: added an alternative method of crafting some burgers, through the correct assembly sequence of modular food.
- tweak: severely cut back on the number of items you can put on burgers, tacos, or kebabs. This had poor design, and things need to be separately resprited by adding them on modular food.

### CODE REVIEW

**Comment by beck-thompson** [comment]
Finally no evil microwave for burgers! 🥳

**Comment by Moomoobeef** [comment]
your work on burgers is making me want to add soups/stews...
perhaps when I get back home (still on vacation rn)

**Comment by KashRas2** [comment]
our elephant 🐘

**Comment by TheShuEd** [comment]
it's ready for review. I think in terms of code it's all ready here. I'll be adding content for a while yet.

File: Content.Shared/Nutrition/Components/FoodMetamorphableByAddingComponent.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+using Content.Shared.Nutrition.EntitySystems;
+
+namespace Content.Shared.Nutrition.Components;
+
+/// <summary>
+/// Attempts to metamorphose a modular food when a new ingredient is added.
+/// </summary>
+[RegisterComponent, Access(typeof(SharedFoodSequenceSystem))]
+public sealed partial class FoodMetamorphableByAddingComponent : Component
```
Comment (metalgearsloth) [inline Content.Shared/Nutrition/Components/FoodMetamorphableByAddingComponent.cs:9]: Needs networking putting in shared does nothin.

File: Content.Shared/Nutrition/Events.cs [hunk 2]
```diff
@@ -67,3 +70,22 @@ public record struct SliceFoodEvent();
 public sealed partial class SliceFoodDoAfterEvent : SimpleDoAfterEvent
 {
 }
+
+/// <summary>
+///    is called when a new ingredient is added to FoodSequence
+/// </summary>
+public sealed class FoodSequenceIngredientAddedEvent : EntityEventArgs
+{
```
Comment (metalgearsloth) [inline Content.Shared/Nutrition/Events.cs:78]: Record struct should also describe what it's raised on.

File: Content.Shared/Nutrition/Prototypes/FoodMetamorphRecipePrototype.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+using Content.Shared.Nutrition.FoodMetamorphRules;
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Nutrition.Prototypes;
+
+/// <summary>
+/// Stores a recipe so that FoodSequence assembled in the right sequence can turn into a special meal.
+/// </summary>
+[Prototype("metamorphRecipe")]
+public sealed partial class FoodMetamorphRecipePrototype : IPrototype
```
Comment (metalgearsloth) [inline Content.Shared/Nutrition/Prototypes/FoodMetamorphRecipePrototype.cs:10]: Datafield string is unnecessary.

File: Content.Shared/Nutrition/Prototypes/FoodMetamorphRecipePrototype.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+{
+    [IdDataField] public string ID { get; private set; } = default!;
+
+    /// <summary>
+    /// The key of the FoodSequence being collected. For example “burger” “taco” etc.
+    /// </summary>
+    [DataField(required: true)]
+    public string Key = string.Empty;
+
```
Comment (metalgearsloth) [inline Content.Shared/Nutrition/Prototypes/FoodMetamorphRecipePrototype.cs:19]: Is this supposed to be a protoid e.g. tags to avoid typos?

**Comment by slarticodefast** [comment]
Oh, I just merged snails, which contain more food sequence elements that need to be adjusted. Sorry. #30765

**Comment by EmoGarbage404** [comment]
Did you pray today
Did you pray today

**Comment by mirrorcult** [comment]
the promised land

**Comment by TheShuEd** [comment]
:0

**Pull request merged**